[지안] step-1 구현

sudago opened the pull request:
안녕하세요 지안입니다!
spring을 다뤄 본 적이 없어서 김영한님 spring 강의를 보며 차근차근 진행하고 있습니다.
따라서 아직 미션을 구현하지 못한 부분이 많고 부족한 부분이 많습니다. 
잘 부탁 드립니다! 😊

\+ 목요일 부터의 진행사항을 reame에 조금 정리해 두었습니다. 

-----
# 기능 구현 목록
## 미션 1단계 
#### 웹 템플릿 엔진 : mustache 사용
- html 중복 제거
- 회원 목록 조회를 위한 반복문 추가
### ✅ 회원 가입
- 회원 가입 페이지에서 회원 정보 입력 후 제출 시 회원 생성
- 회원 생성 후 home으로 redirect
### ✅ 회원 목록 조회
- home에서 회원 목록 아이콘 클릭 시 회원 목록 페이지로 이동
- 회원 목록을 출력
### 회원 프로필 조회 (진행중)
- 회원 목록에서 회원 클릭 시 개별 회원의 프로필 정보 출력 (진행중)

## 미션 2단계 (미구현)
- WebMVCConfigurer 적용하기
### 글쓰기
### 글 목록 조회 기능
### 게시글 상세보기
### 추가 요구 사항 (선택)
#### 회원 정보 수정
- PUT method 사용


-----

# 각 기능에 따른 URL과 메서드 컨벤션

url | 기능
-- | --
GET /todos | List all todos
POST /todos | Create a new todo
GET /todos/:id | Get a todo
PUT /todos/:id | Update a todo
DELETE /todos/:id | Delete a todo and its items
GET /todos/:id/items | Get a todo item
PUT /todos/:id/items | Update a todo item
DELETE /todos/:id/items | Delete a todo item

<br class="Apple-interchange-newline">

sudago committed:
학습 계획 작성 및 폴더 이동

sudago committed:
docs: gradle 설정 추가 및 gitignore 추가

sudago committed:
style: 폴더 이동에 따른 경로 자동 수정

sudago committed:
docs: 한 일 수정

sudago committed:
style: mustache로 html 형식 수정

sudago committed:
style: mustache로 html 형식 수정

sudago committed:
feat: 회원가입 버튼 클릭 시 페이지로 이동 기능 추가

Controller로 연결해서 페이지를 이동시켜보았다.

sudago committed:
docs: properties 파일 설정

sudago committed:
feat: 연습을 위한 컨트롤러 추가

sudago committed:
feat: User를 저장하기 위한 클래스 구현

sudago committed:
test: 테스트 코드 작성

sudago commented on src/main/resources/static/index.html:
```diff
@@ -65,8 +65,8 @@
-                <li><a href="user/login.html" role="button">로그인</a></li>
-                <li><a href="user/form.html" role="button">회원가입</a></li>
+                <li><a href="../templates/user/login.html" role="button">로그인</a></li>
+                <li><a href="/user/create" role="button">회원가입</a></li>
```
원본 index를 고쳐도 되는건지? 의문입니다. 경로를 수정하지 않아도 template으로 갈 수 있는 방법이 따로 있는건지 궁금합니다.

sudago commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,32 @@
+        return "/user/form"; // template 폴더에 만들었다면
+    }
+
+    @PostMapping("/user/create")
```
GetMapping쪽은 잘 동작하는데, PostMapping쪽이 잘 동작하지 않아서 문제를 찾는 중입니다.

sudago commented on src/main/resources/templates/layout/header.html:
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
```
mustache를 써서 반복되는 부분은 따로 layout 폴더 안에 분리해서 두었습니다. 그런데 \<html> 태그까지 묶어버려도 괜찮은지 궁금합니다. 잘 동작하기는 하는데 이렇게 묶는 방식이 올바른 방법인지 궁금합니다.

jinan159 replied:
p2)
이 부분은 잘 동작만 한다면 크게 고민하지 않아도 되는 영역이라 생각합니다.
브라우저에서 접속하여 개발자도구를 보고 어떻게 HTML이 생성되었는지 한번 확인해보시면 좋을 것 같습니다.

p3)
그리고 시간 여유가 있으시다면 mustache와 같은 템플릿 엔진의 동작원리를 한번 공부해보시면 도움이 되실 것 같습니다.
(긴 시간을 들일 필요는 없습니다.)

sudago committed:
docs: gradle 설정 추가

sudago committed:
docs: 할 일 목록 작성

sudago committed:
style: 폴더 이동


style: 폴더 이동

sudago committed:
style: 회원 번호와 회원id 구분을 위한 변수 변경

sudago committed:
feat: 회원 가입 기능 추가

User 객체와 MemoryUserRepository에 저장

sudago committed:
feat: home("/") URL 설정

sudago committed:
feat: 회원 목록 페이지 기능 구현

sudago committed:
style: 폴더 이동

sudago committed:
style: 주석 추가

sudago commented on src/main/java/kr/codesqaud/cafe/controller/HomeController.java:
```diff
@@ -0,0 +1,17 @@
+
+@Controller
+public class HomeController {
+    @GetMapping("/")
+    public String home(){
+        return "/index";
+    }
+
+    @GetMapping("/index")
+    public String index(){
+        return "/index";
+    }
```
동일한 url을 return 하는 메서드가 두개 있어도 되는지 궁금합니다. (같은 일을 하는 메서드가 둘이나 있어도 되는건지, 아니면 각 html에서 해당하는 부분을 직접 찾아서 수정해야하는건지 궁금합니다.)

jinan159 replied:
p2) 
위쪽 `home()` 메서드를 지워도 index 페이지로 이동할 수 있습니다.
왜 이렇게 되는지 한번 고민해보시는걸 추천드립니다.
(힌트 키워드는 web server default page 입니다.)

그리고 지안님이 의문을 가지신 것 처럼 보통 같은일을 하는 메서드는 두지 않는것이 일반적입니다.

sudago commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,51 @@
+        this.userService = userService;
+    }
+
+    @GetMapping("/user/create")
+    public String createForm(){
+        return "/user/form"; // template 폴더에 만들었다면
```
받는 url(/user/create) 와 return 하는 (/user/form) 은 일치하지 않아도 되나요? 나중에 오류가 났을 때, URL 이름이 일치하지 않으면 찾기가 힘들 것 같아 궁금합니다.

jinan159 replied:
URL과 어떤 VIEW의 이름을 통일할 필요는 없습니다.
그리고 오류가 나더라도, 사실 로깅을 잘해놓으면 어떤 파일의 어떤 라인에서 발생했는지까지 알 수 있기 때문에 찾기 어려운 부분은 걱정하지 않으셔도 될거같아요.

sudago commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,51 @@
+        return "/user/list";
+    }
+
+    @GetMapping("/users/profile") // ("/users/{{userId}}")
```
이 부분은 아직 구현중입니다.

sudago commented on src/main/java/kr/codesqaud/cafe/domain/User.java:
```diff
@@ -0,0 +1,54 @@
+        // 회원가입 할 때 전달한 값을 저장할 수 있는 필드를 생성한 후 setter와 getter 메소드를 생성한다.
+    // 사용자 목록을 관리하기 위해 JCF (Java Collections Framework) 를 사용한다.
+
+    private Long number; // 회원번호
```
회원 번호 변수명이 명확하지 않은 것 같은데 Jay는 어떻게 생각하시는지 궁금합니다.

sudago replied:
회원 번호 변수명이 명확하지 않은 것 같은데 Jay는 어떻게 생각하시는지 궁금합니다. 또한, User가 회원번호를 가지고 있는게 올바른지 궁금합니다.

jinan159 commented:
지안님 반갑습니다. 앞으로 5주간 지안님의 미션을 리뷰하게된 Jay 입니다.
리뷰를 하면서는 앞서 제가 피드백을 드리는 내용의 중요성이나 의도를 더 명확하게 표현하기 위해 Pn룰을 적용할 예정입니다.
- [코드 리뷰 in 뱅크샐러드 개발 문화 - 커뮤니케이션 비용을 줄이기 위한 Pn 룰](https://blog.banksalad.com/tech/banksalad-code-review-culture/#%EC%BB%A4%EB%AE%A4%EB%8B%88%EC%BC%80%EC%9D%B4%EC%85%98-%EB%B9%84%EC%9A%A9%EC%9D%84-%EC%A4%84%EC%9D%B4%EA%B8%B0-%EC%9C%84%ED%95%9C-pn-%EB%A3%B0)

글에서는 5단계로 되어있는데, 저는 아래처럼 3단계로 변형하여 사용할 예정이니 참고 부탁드립니다.

- P1: 꼭 반영해주세요 (Request changes)
  - 중대한 오류, 버그 가능성을 내포하고 있는 경우
  - 리뷰 요청자는 요청을 반영하거나, 혹은 반영할 수 없는 이유를 설명해야합니다.

- P2: 웬만하면 반영해 주세요 (Comment)
  - 오류는 아니지만 성능 혹은 구조적으로 문제가될 수 있는 경우
  - 리뷰 요청자는 요청을 반영하거나, 혹은 반영할 수 없는 이유를 설명해야합니다.

P3: 반영해도 좋고 넘어가도 좋습니다 (Approve)
  - 리뷰어의 사소한 의견으로 이에 대해 답변을 달아도 되고, 안달아도 괜찮습니다.
    (하지만 확인했는지 여부를 확인하기 위해 가능한 이모지는 달아주시면 감사하겠습니다.)

단순 답변은 Pn 태그를 달지 않을 예정이며, PR은 일요일 자정에 제가 머지할 예정입니다.
(`Create a merge commit` 으로 할 예정)

jinan159 commented on src/main/java/kr/codesqaud/cafe/domain/User.java:
```diff
@@ -0,0 +1,54 @@
+        // 회원가입 할 때 전달한 값을 저장할 수 있는 필드를 생성한 후 setter와 getter 메소드를 생성한다.
+    // 사용자 목록을 관리하기 위해 JCF (Java Collections Framework) 를 사용한다.
+
+    private Long number; // 회원번호
```
> 회원 번호 변수명이 명확하지 않은 것 같은데 Jay는 어떻게 생각하시는지 궁금합니다. 또한, User가 회원번호를 가지고 있는게 올바른지 궁금합니다.

일단 회원번호라는건 유저를 식별하기 위한 속성일텐데, number라는 단어는 확 와닿지는 않는 것 같습니다.

p3)
이름을 지을때 한글 명칭 <--> 영어로 직역하는것도 좋지만, 해당 변수의 역할이 뭔지 생각해서 그에 맞는 이름을 지어주는것도 하나의 방법입니다.
그리고 저는 이름 정하기가 애매할때 필드만 보기 보다는 이를 불러서 사용할때의 관점에서 생각보는 편입니다.
User user = new User()l
user.~~~();
그런데 사실 이렇게해도 좋은 이름이 딱 떠오르지 않을 수 있습니다.
명칭의 경우에는 다른사람들이 일반적으로 쓰는 명칭이 어떤게 있는지 한번 참고해보는것도 좋습니다.

그리고 회원번호가 User를 식별하기 위한 고유 번호라면, 저는 User가 갖고있는건 저는 지극히 정상적이라고 생각됩니다.

jinan159 commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,51 @@
+@Controller
+public class UserController {
+    private final UserService userService;
+    @Autowired
```
p2)
빈의 주입방식은 여러 방식이 있는데요. 종류와 각 방식에 대해 간단히 설명 부탁드려도 될까요?
그리고 여기 17번 라인의 `@Autowired` 어노테이션이 꼭 필요한지 한번 살펴보시는걸 추천드립니다

sudago replied:
사실 강의를 보고 그대로 따라한 거라서 특별한 의미는 없습니다. @Autowrited 가 무엇인지 정확히 알아보고 사용하도록 하겠습니다. 감사합니다!

jinan159 commented on src/main/java/kr/codesqaud/cafe/repository/MemoryUserRepository.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Override
+    public User save(User user) {
+        user.setNumber(++sequence); //
```
p1)
혹시 회원가입 요청이 아주 많이 들어오는 경우에 sequence 를 증가하는 부분에서 동시성 문제가 일어날 수 있지 않을까요?
그리고 이를 해결할 수 있는 방법이 있는데 한번 알아보시면 좋을 것 같습니다.
(힌트 키워드는 Atomic 입니다)

sudago replied:
감사합니다! 추가로 알아보도록 하겠습니다!

jinan159 commented on src/main/java/kr/codesqaud/cafe/repository/MemoryUserRepository.java:
```diff
@@ -0,0 +1,41 @@
+@Repository
+public class MemoryUserRepository implements UserRepository {
+    // 원래는 아래의 store, sequence는 동시성 문제때문에 concurrentHashMap 등을 쓴다.
+    private static Map<Long, User> store = new HashMap<>();
```
p1)
이 부분은 동시성 문제가 있다고 말씀해주셨는데, 왜 동시성 문제가 일어날 수 있는 일반 HashMap 을 선언하셨나요?

sudago replied:
이 부분 역시 강의를 듣고 따라한 부분인데, 전체적인 흐름을 이해하고 나서, 보다 어려운 개념인 currentHashMap을 사용하도록 수정하는 것이 좋을 것 같아서 일단 HashMap으로 선언했습니다. 주말동안 추가로 학습해보도록 하겠습니다!

jinan159 commented on src/main/java/kr/codesqaud/cafe/repository/MemoryUserRepository.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Override
+    public List<User> findAll() { // 모든 user가 반환됨.
+        return new ArrayList<>(store.values());
```
p2) 
`ArrayList`로 반환하면 변경 가능한 리스트인데요.
혹시 변경 불가능한 리스트로 반환하는 방법이 있을까요? 
있다면 어떤 방법이 있을지 한번 찾아보시면 좋을 것 같습니다.

sudago replied:
네 추가로 학습해보도록 하겠습니다!

jinan159 commented on src/main/java/kr/codesqaud/cafe/service/UserService.java:
```diff
@@ -0,0 +1,45 @@
+
+    private final UserRepository userRepository;
+
+    @Autowired
```
p2)
이 부분도 `UserController`에 달았던 코멘트와 마찬가지로 한번 고민 해보시면 좋을 것 같습니다.

jinan159 commented on src/main/java/kr/codesqaud/cafe/service/UserService.java:
```diff
@@ -0,0 +1,45 @@
+    private void validateDuplicateUser(User user) {
+        // ifPresent = 값이 있으면. optional으로 감쌌기 때문에 가능하다.
+        userRepository.findByName(user.getName())
+                .ifPresent(u -> {
+                    throw new IllegalStateException("이미 존재하는 회원입니다.");
+                });
```
p3)
여기서 IllegalStateException을 사용하신건 적절한 선택인 것 같습니다.
그런데 조금 더 욕심을 내보자면, 이보다 더 명확안 예외를 발생시킬 방법이 있을까요?
한번 고민 해봐주시면 좋을 것 같습니다.

sudago replied:
예외처리에 아직 미숙한 부분이 많아서 추가로 공부해보겠습니다!

jinan159 commented on src/test/java/kr/codesqaud/cafe/repository/MemoryUserRepositoryTest.java:
```diff
@@ -0,0 +1,62 @@
+
+import static org.assertj.core.api.Assertions.*;
+
+public class MemoryUserRepositoryTest {
```
테스트를 작성해주신 부분 아주 좋습니다.
각 메서드의 성공케이스에 대한 적절한 테스트를 작성하신 것 같네요.
다만 여기서 조금 더 욕심을 내자면, 몇가지 드릴 말씀이 있습니다.

p2)
먼저, "테스트를 할 때 구현(implementation)이 아니라 설계(interface)에 맞춰야 한다" 입니다.
관련 내용은 아래 링크를 첨부할테니 읽어보시는걸 추천드립니다.
https://github.com/david-learner/java-study/tree/master/2018OKKYCON#%EC%84%A4%EA%B3%84

그리고 성공케이스 뿐만 아니라, 실패 케이스에 대한 테스트도 중요합니다.
예를 들면, save 메서드는 중복된 회원번호를 입력하는 경우 예외가 발생하죠.
이게 잘 동작하는지 검증하는 케이스도 있으면 더 좋을 것 같네요.
다른 메서드들도 한번 생각해봐주시면 좋을 것 같습니다.

sudago replied:
오 감사합니다! 재미있을 것 같은 학습 키워드네요ㅎㅎ

jinan159 commented on src/test/java/kr/codesqaud/cafe/service/UserServiceTest.java:
```diff
@@ -0,0 +1,75 @@
+        IllegalStateException e = assertThrows(IllegalStateException.class, () -> userService.join(user2));
+
+        assertThat(e.getMessage()).isEqualTo("이미 존재하는 회원입니다.");
+/*
+        try {
+            userService.join(user2);
+            fail();
+        } catch (IllegalStateException e) {
+            assertThat(e.getMessage()).isEqualTo("이미 존재하는 회원입니다.");
+        }
+
+ */
+
+        // then
```
p2)
불필요한 주석이라면 지워주시는게 좋습니다.

jinan159 commented on src/test/java/kr/codesqaud/cafe/service/UserServiceTest.java:
```diff
@@ -0,0 +1,75 @@
+        userService.join(user1);
+        IllegalStateException e = assertThrows(IllegalStateException.class, () -> userService.join(user2));
+
+        assertThat(e.getMessage()).isEqualTo("이미 존재하는 회원입니다.");
```
p3)
given - when - then 방식의 테스트 작성 레이아웃을 사용하신 것 같네요.
여기서 then 주석의 위치를 확인해주시면 좋을 것 같아요.

jinan159 commented on src/test/java/kr/codesqaud/cafe/service/UserServiceTest.java:
```diff
@@ -0,0 +1,75 @@
+        // then
+    }
+
+    @Test
+    void findUsers() {
+    }
+
+    @Test
+    void findOne() {
+    }
```
p3)
이 부분은 시간이 부족하여 작성하지 못한 테스트인가요?

sudago replied:
이 부분 역시 강의를 따라 한 내용인데, 현재 제가 구현하고 있는 부분과는 일치하지 않아서 제거하도록 하겠습니다!

jinan159 reviewed (changes requested):
몇가지 요청드린 사항이 있는데, 확인 해주시면 좋을 것 같습니다.
만약 확인이 덜 되더라도 일요일 자정에 머지할 예정이지만, 혹시 추가 의견이 있으신 경우 댓글 달아주시면 확인하고 답변 드리겠습니다.

sudago commented:
제가 spring에 대한 이해도가 많이 부족한 상황에서 Jay님이 리뷰해주실 부분이 있을까 걱정했었는데, 좋은 학습 키워드들을 많이 알려주셔서 감사합니다.
피드백 주신 부분을 더 꼼꼼히 확인하고 싶어서 답변을 미룬 부분이 있었는데, 피드백 주신 것을 확인하자마자 답변 드리는게 더 좋았을 것 같다는 생각이 듭니다.
동료들과 소통하면서도 궁금점이 해소되지 않았던 부분들이 있었는데, Jay님이 제가 생각치도 못했던 부분까지 말씀해주셔서 정말 좋았습니다!
머지 해주셔도 괜찮을 것 같습니다. 학습이 필요한 부분은 차근차근 공부해나가도록 하겠습니다!
앞으로도 잘 부탁드리겠습니다! 🙇🏻‍♀️

jinan159 commented:
빠른 확인 감사드려요.
다만 p1로 된 부분은 리뷰 드린 내용을 반영해주시면 좋을 것 같습니다.

수정해서 그대로 `sudago:feature1-1` 브랜치로 푸시 해주시면 됩니다.

sudago commented:
네 알겠습니다! 일요일 자정 전까지 p1 부분 전부 수정하겠습니다 😊

sudago committed:
style: 회원 번호 변수명 변경

회원 번호의 변수명을 직관적으로 표시하기 위해 변경

sudago committed:
fix: 동시성 문제 고려해서 concurrentHashMap으로 변경

concurrentHashMap이 전체적인 틀을 학습하기에 어려울 것 이라고 생각해서 hashMap을 사용했는데, 일단 동시성 문제를 고려해야 하는게 맞다고 판단해서 변경.

sudago committed:
fix: 회원 번호를 AtomicLong 사용해 자동으로 부여하는 것으로 변경

직접 sequence를 증가시켜 각 회원이 생성될 때 마다 회원번호를 부여하는 것으로 했었는데, 회원 가입 요청 또한 동시성 문제가 발생 할 수 있으니 이를 해결 할 수 있는 AtomicLong으로 변경함.

sudago commented:
p1 부분 전부 수정했습니다. 나머지 부분도 월요일부터 학습해서 수정하도록 하겠습니다! 감사합니다!

jinan159 commented:
학습하시면서 실제 기능을 구현해 나가며 이런 코드나 이런 구조가 맞는지, 현업에서도 어떻게 하는지 궁금하실거라 생각합니다.
하지만, 지금 단계에서 중요한건 `내 생각`이 들어간 코드입니다.
강의에서 배우신 구조, 코드, 지식들을 그냥 받아들이시가 보다는 의문을 품고 스스로 해결하는 과정에서 더 많이 배우실거라 생각합니다.
저는 개발에서는 '정답'이란 존재하지 않고 '최선책'만 존재한다고 생각해요.
그런데 내가 믿던 '최선책'보다 더 좋은 선택지를 찾아내기 위해서는 지식들과 더불어 정리된 '내 생각'이 있어야합니다.
학습과정에서 지식과 함께 지안님만의 '생각'들을 잘 쌓아나가시길 응원합니다.

이번주도 고생하셨습니다~!

PR은 일요일 자정에 머지한다고 했는데, 지금 머지해도될 것 같아서 머지하겠습니다.

sudago commented:
이번주는 조언해주신대로 강의보다는 스스로 생각해서 짜보는 연습을 해보도록 하겠습니다! 
좋은 피드백 감사합니다~ 이번주도 잘 부탁드리겠습니다ㅎㅎ
